types/underscore/index.d.ts

```diff
@@ -3436,36 +3438,27 @@ declare module _ {
         values(object: any): any[];
 
         /**
-         * Like map, but for objects. Transform the value of each property in turn.
-         * @param object The object to transform
-         * @param iteratee The function that transforms property values
-         * @param context The optional context (value of `this`) to bind to
-         * @return a new _.Dictionary of property values
-         */
-        mapObject<T, U>(object: _.Dictionary<T>, iteratee: (val: T, key: string, object: _.Dictionary<T>) => U, context?: any): _.Dictionary<U>;
-
-        /**
-         * Like map, but for objects. Transform the value of each property in turn.
-         * @param object The object to transform
-         * @param iteratee The function that tranforms property values
-         * @param context The optional context (value of `this`) to bind to
-         */
-        mapObject<T>(object: any, iteratee: (val: any, key: string, object: any) => T, context?: any): _.Dictionary<T>;
-
-        /**
-         * Like map, but for objects. Retrieves a property from each entry in the object, as if by _.property
-         * @param object The object to transform
-         * @param iteratee The property name to retrieve
-         * @param context The optional context (value of `this`) to bind to
+         * Like map, but for objects. Transform the value of each property in
+         * turn.
+         * @param object The object to transform.
+         * @param iteratee The iteratee to use to transform property values.
+         * @param context `this` object in `iteratee`, optional.
+         * @returns A new object with all of `object`'s property values
+         * transformed through `iteratee`.
          */
-        mapObject(object: any, iteratee: string, context?: any): _.Dictionary<any>;
+        mapObject<V extends object, I extends Iteratee<V, any, TypeOfDictionary<V, any>>>(
+            object: V,
+            iteratee: I,
+            context?: any
+        ): { [K in keyof V]: IterateeResult<I, V[K], EnumerableKey[], undefined> };
```

Comment:
Hmm, I was thinking that this was a little different than checking the type of a collection, but I suppose union types can happen both in collections and on individual properties, and I suppose in the latter case `undefined` would be an inaccurate type to wind up with. I'll try some stuff quick, but I might end up dropping these funky bits altogether since I'm not sure that I can reliably end up with a final result of `undefined` in all cases (though I'd admittedly prefer to end up with `unknown` instead of `any` where I can't figure out the type and the likely actual type is `undefined`).

Reply:
Well, with that realization `Iteratee` and `IterateeResult` end up having almost no changes at all.